README.markdown
# unix_utils

Like FileUtils, but provides zip, unzip, bzip2, bunzip2, tar, untar, sed, du, md5sum, shasum, cut, head, tail, wc, unix2dos, dos2unix, iconv, curl, perl, etc.

You must have these binaries in your `PATH`. _Not_ a pure-ruby implementation of all these UNIX greats!

Works in MRI 1.8.7+, MRI 1.9.2+, and JRuby 1.6.7+. No gem dependencies; uses stdlib

## Real-world usage

<p><a href="http://brighterplanet.com"><img src="https://s3.amazonaws.com/static.brighterplanet.com/assets/logos/flush-left/inline/green/rasterized/brighter_planet-160-transparent.png" alt="Brighter Planet logo"/></a></p>

We use `unix_utils` for [data science at Brighter Planet](http://brighterplanet.com/research) and in production at

* [Brighter Planet's impact estimate web service](http://impact.brighterplanet.com)
* [Brighter Planet's reference data web service](http://data.brighterplanet.com)

Originally extracted from [`remote_table`](https://github.com/seamusabshere/remote_table)

## Philosophy

Use a subprocess to perform a big task and then get out of memory.

<table>
  <tr>
    <td rowspan="2"><img src="https://github.com/seamusabshere/unix_utils/raw/master/unix-philosophy-cover.png" alt="cover of the Unix Philosophy book" /></td>
    <td><img src="https://github.com/seamusabshere/unix_utils/raw/master/unix-philosophy-quote-pg1.png" alt="Tenet 2: Make Each Program Do One Thing Well. The best programs, like Cousteau's lake fly, does but one task in its life and does it well. The program is loaded into memory, accomplishes its function, and then gets out ot the way to allow" /></td>
  </tr>
  <tr>
    <td><img src="https://github.com/seamusabshere/unix_utils/raw/master/unix-philosophy-quote-pg2.png" alt="the next single-minded program to begin. This sounds simple, yet it may surprise you how many software developers have difficulty sticking to this singular goal." /></td>
  </tr>
</table>

## Rules (what you can expect)

For commands like zip, untar, sed, head, cut, dos2unix, etc.:

1. Just returns a path to the output, randomly named, located in the system tmp dir (`UnixUtils.unzip('kittens.zip)` &rarr; `'/tmp/unix_utils-129392301-kittens'`)
2. Never touches the input
3. Sticks a useful file extension on the output, if applicable (`UnixUtils.tar('puppies/')` &rarr; `'/tmp/unix_utils-99293192-puppies.tar'`)

For commands like du, md5sum, shasum, etc.:

1. Just returns the good stuff (the checksum, for example, not the filename that is listed after it in the standard command output)
2. Never touches the input

## But I can just spawn these myself

This lib was created to ease the pain of remembering command options for Gentoo, deciding which spawning method to use, possibly handling pipes...

    require 'tmpdir'
    destdir = File.join(Dir.tmpdir, "kittens_#{Kernel.rand(1e11)}")
    require 'open3'
    Open3.popen3('unzip', '-q', '-n', 'kittens.zip, '-d', destdir) do |stdin, stdout, stderr|
      stdin.close
      @error_message = stderr.read
    end

is replaced safely with

    destdir = UnixUtils.unzip 'kittens.zip'

## But I can just use `Digest::SHA256`

(Note: [Balazs Kutil](https://github.com/bkutil) pointed out [this is a bad example](https://gist.github.com/1950707)... I will replace it soon)

This will load an entire file into memory before it can be processed...

    require 'digest'
    str = Digest::SHA256.hexdigest File.read('kittens.zip')

... so you're really replacing this ...

    sha256 = Digest::SHA256.new
    File.open('kittens.zip', 'r') do |f|
      while chunk = f.read(4_194_304)
        sha256 << chunk
      end
    end
    str = sha256.hexdigest

You get the same low memory footprint with

    str = UnixUtils.shasum 'kittens.zip', 256

## Compatibility

Uses `open3` because it's in the Ruby stdlib and is consistent across MRI and JRuby.

## Wishlist

* cheat sheet based on [GNU Coreutils cheat sheet](www.catonmat.net/download/gnu-coreutils-cheat-sheet.pdf)
* yarddocs
* properly use Dir.tmpdir(name), etc.
* smarter tmp file name generation - don't include url params for curl, etc.

## Authors

* Seamus Abshere <seamus@abshere.net>

## Copyright

Copyright (c) 2012 Seamus Abshere


CHANGELOG
unreleased

* Enhancements

  * Clean up generation of temp file names

0.0.15 / 2012-11-09

* Enhancements

  * Back to stdlib open3 because C extension support is not enabled by default in JRuby 1.7+

* Known issues

  * stderr is not captured in JRuby 1.7.0 because of https://jira.codehaus.org/browse/JRUBY-6986

0.0.14 / 2012-07-06

* Breaking changes

  * UnixUtils.du_sk -> UnixUtils.du (but it's still `du -sk`)
  * Heuristic for detecting attempts to curl a local file changed - now it looks for absence of '://'

0.0.13 / 2012-05-16

* Bug fixes

  * `which -s` didn't work on all platforms

0.0.12 / 2012-05-09

* Bug fixes

  * Unreleased code (dos2unix and unix2dos) appeared in gem release - got rid of that.

0.0.11 / 2012-05-07

* Bug fixes

  * Fix handling of local files by UnixUtils.curl

0.0.10 / 2012-04-23

* Breaking changes

  * Added UnixUtils.du_sk, which always returns bytes; removed UnixUtils.du, which returned block counts, which I doubt you wanted.

* Enhancements

  * If you call a method with _s at the end, you get a string back. For example: UnixUtils.head_s
  * UnixUtils.tmp_path no longer junks up tmp paths with lots of "unix_utils_RANDOMSTRING"

0.0.9 / 2012-04-12

* Bug fixes

  * Correct implementation of :read_from

0.0.8 / 2012-04-12

* Bug fixes

  * Let perl handle its own input files (ARGF)... my implementation using IO.select was not working.

0.0.7 / 2012-04-11

* Bug fixes

  * Make sure UnixUtils.tmp_path doesn't create too-long filenames

0.0.6 / 2012-04-11

* Enhancements

  * iconv command silently discards invalid characters

0.0.5 / 2012-03-21

* Enhancements

  * Change pipe buffer size to 64kb (it was 4mb before), which seems to make everything run faster

* Bug fixes

  * Thanks to @leomao10, make sed work more reliably on large files by using its input-file argument rather than piping in data

0.0.4 / 2012-03-19

* Bug fixes

  * NoMethodError when trying to print $stderr - thanks @leomao10 !

0.0.3 / 2012-03-19

* Enhancements

  * Uses https://github.com/rtomayko/posix-spawn to make things faster - suggested by @jjb in https://github.com/seamusabshere/unix_utils/issues/1
  * Depends less on chdir to work properly - expands file paths immediately

* Bug fixes

  * Correctly use pipes (I hope) - imitate POSIX::Spawn::Child's use of IO.select

0.0.2 / 2012-02-16

* Bug fixes

  * Fix use of splat internally so that it actually runs in MRI 1.8.7 and JRuby

0.0.1 / 2012-02-16 (yanked!)

* Birthday!


test/helper.rb
require 'rubygems'
require 'bundler/setup'
require 'pry'

require 'minitest/spec'
require 'minitest/autorun'
if RUBY_VERSION >= '1.9'
  require 'minitest/reporters'
  MiniTest::Reporters.use! MiniTest::Reporters::SpecReporter.new
end

require 'unix_utils'

require 'stringio'
require 'fileutils'
require 'tmpdir'
require 'tempfile'

module TestHelper
  def assert_does_not_touch(method_id, *args)
    infile_or_srcdir = args.first
    mtime = File.mtime infile_or_srcdir
    kind = File.file?(infile_or_srcdir) ? :file : :directory
    case kind
    when :file
      checksum = UnixUtils.shasum infile_or_srcdir, 256
    when :directory
      size = UnixUtils.du infile_or_srcdir
    end
    destdir = UnixUtils.send(*([method_id] + args))
    safe_delete destdir
    File.mtime(infile_or_srcdir).must_equal mtime
    case kind
    when :file
      UnixUtils.shasum(infile_or_srcdir, 256).must_equal checksum
    when :directory
      UnixUtils.du(infile_or_srcdir).must_equal size
    end
  end

  def assert_unpack_dir(method_id, infile)
    destdir = UnixUtils.send method_id, infile
    File.directory?(destdir).must_equal true
    Dir.entries(destdir).must_equal %w{ . .. hello_world.txt hello_world.xml }
    File.dirname(destdir).start_with?(Dir.tmpdir).must_equal true
    safe_delete destdir
  end

  def assert_unpack_file(method_id, infile)
    outfile = UnixUtils.send method_id, infile
    File.file?(outfile).must_equal true
    `file #{outfile}`.chomp.must_match %r{text}
    File.dirname(outfile).start_with?(Dir.tmpdir).must_equal true
    safe_delete outfile
  end

  def assert_pack(method_id, infile)
    outfile = UnixUtils.send method_id, infile
    File.file?(outfile).must_equal true
    `file #{outfile}`.chomp.must_match %r{\b#{method_id.to_s.downcase}\b}
    File.dirname(outfile).start_with?(Dir.tmpdir).must_equal true
    safe_delete outfile
  end

  def safe_delete(path)
    path = File.expand_path path
    raise "Refusing to rm -rf #{path} because it's not in #{Dir.tmpdir}" unless File.dirname(path).start_with?(Dir.tmpdir)
    FileUtils.rm_rf path
  end
end

class MiniTest::Spec
  include TestHelper
end


test/test_unix_utils.rb
# encoding: UTF-8
require 'helper'

describe UnixUtils do
  before do
    @old_pwd = Dir.pwd
    Dir.chdir File.expand_path('../target', __FILE__)
  end

  after do
    Dir.chdir @old_pwd
  end

  describe "errors" do
    it "gets printed to stderr" do
      begin
        old_stderr = $stderr
        capture = StringIO.new
        $stderr = capture
        UnixUtils.unzip(__FILE__)
        capture.rewind
        capture.read.must_match %r{End-of-central-directory signature not found}
      ensure
        $stderr = old_stderr
      end
    end
  end

  describe :curl do
    it "downloads to a temp file" do
      outfile = UnixUtils.curl('http://brighterplanet.com')
      File.read(outfile).must_match %r{sustain}i
      safe_delete outfile
    end

    it "deals safely with local files" do
      UnixUtils.curl_s('utf8.txt').must_include 'Hola, ¿cómo estás?'
      UnixUtils.curl_s(File.expand_path('utf8.txt')).must_include 'Hola, ¿cómo estás?'
      UnixUtils.curl_s("file://#{File.expand_path('utf8.txt')}").must_include 'Hola, ¿cómo estás?'
    end
  end

  describe :shasum do
    it "checksums a file with SHA-1" do
      UnixUtils.shasum('directory.zip', 1).must_equal 'c0abb36c923ed7bf87ebb8d7097cb8e264e528d2'
    end
    it "checksums a file with SHA-256" do
      UnixUtils.shasum('directory.zip', 256).must_equal '661af2b7b0993088263228b071b649a88d82a6a655562162c32307d1e127f27a'
    end
    it "works just as well with absolute paths" do
      target_path = File.join(Dir.pwd, 'directory.zip')
      Dir.chdir('/') do
        UnixUtils.shasum(target_path, 256).must_equal '661af2b7b0993088263228b071b649a88d82a6a655562162c32307d1e127f27a'
      end
    end
  end

  describe :md5sum do
    it "checksums a file" do
      UnixUtils.md5sum('directory.zip').must_equal 'd6e15da798ae19551da6c49ec09afaef'
    end
  end

  describe :du do
    it "calculates the size of a directory in bytes" do
      UnixUtils.du('directory').must_equal 8
    end
  end

  describe :unzip do
    before do
      @infile = 'directory.zip'
      @anonymous_infile = 'directory-really-a-z_i_p-shh'
    end
    it "unpacks a DIRECTORY located in the tmp directory" do
      assert_unpack_dir :unzip, @infile
    end
    it "accepts unsemantic filenames" do
      assert_unpack_dir :unzip, @anonymous_infile
    end
    it "does not touch the infile" do
      assert_does_not_touch :unzip, @infile
    end
  end

  describe :untar do
    before do
      @infile = 'directory.tar'
      @anonymous_infile = 'directory-really-a-t_a_r-shh'
    end
    it "unpacks a DIRECTORY located in the tmp directory" do
      assert_unpack_dir :untar, @infile
    end
    it "accepts unsemantic filenames" do
      assert_unpack_dir :untar, @anonymous_infile
    end
    it "does not touch the infile" do
      assert_does_not_touch :untar, @infile
    end
  end

  describe :bunzip2 do
    before do
      @infile = 'file.bz2'
      @anonymous_infile = 'file-really-a-b_z_2-shh'
    end
    it "unpacks a FILE located in the tmp directory" do
      assert_unpack_file :bunzip2, @infile
    end
    it "accepts unsemantic filenames" do
      assert_unpack_file :bunzip2, @anonymous_infile
    end
    it "does not touch the infile" do
      assert_does_not_touch :bunzip2, @infile
    end
  end

  describe :gunzip do
    before do
      @infile = 'file.gz'
      @anonymous_infile = 'file-really-a-g_z-shh'
    end
    it "unpacks a FILE located in the tmp directory" do
      assert_unpack_file :gunzip, @infile
    end
    it "accepts unsemantic filenames" do
      assert_unpack_file :gunzip, @anonymous_infile
    end
    it "does not touch the infile" do
      assert_does_not_touch :gunzip, @infile
    end
  end

  describe :bzip2 do
    before do
      @infile = 'directory.tar'
    end
    it "packs a FILE to a FILE in the tmp directory" do
      assert_pack :bzip2, @infile
    end
    it "does not touch the infile" do
      assert_does_not_touch :bzip2, @infile
    end
    it "sticks on a useful extension" do
      outfile = UnixUtils.bzip2 @infile
      File.extname(outfile).must_equal '.bz2'
      safe_delete outfile
    end
  end

  describe :gzip do
    before do
      @infile = 'directory.tar'
    end
    it "packs a FILE to a FILE in the tmp directory" do
      assert_pack :gzip, @infile
    end
    it "does not touch the infile" do
      assert_does_not_touch :gzip, @infile
    end
    it "sticks on a useful extension" do
      outfile = UnixUtils.gzip @infile
      File.extname(outfile).must_equal '.gz'
      safe_delete outfile
    end
  end

  describe :zip do
    before do
      @srcdir = 'directory'
    end
    it "packs a DIRECTORY to a FILE in the tmp directory" do
      assert_pack :zip, @srcdir
    end
    it "does not touch the infile" do
      assert_does_not_touch :zip, @srcdir
    end
    it "sticks on a useful extension" do
      outfile = UnixUtils.zip @srcdir
      File.extname(outfile).must_equal '.zip'
      safe_delete outfile
    end
  end

  describe :tar do
    before do
      @srcdir = 'directory'
    end
    it "packs a DIRECTORY to a FILE in the tmp directory" do
      assert_pack :tar, @srcdir
    end
    it "does not touch the infile" do
      assert_does_not_touch :tar, @srcdir
    end
    it "sticks on a useful extension" do
      outfile = UnixUtils.tar @srcdir
      File.extname(outfile).must_equal '.tar'
      safe_delete outfile
    end
  end

  describe :perl do
    before do
      @f = Tempfile.new('perl.txt')
      @f.write "badWord\n"*10_000
      @f.flush
      @infile = @f.path
    end
    after do
      @f.close
    end
    it "processes a file" do
      outfile = UnixUtils.perl(@infile, 's/bad/good/g')
      File.read(outfile).must_equal "goodWord\n"*10_000
      safe_delete outfile
    end
    it "does not touch the infile" do
      assert_does_not_touch :perl, @infile, 's/bad/good/g'
    end
    it "keeps the original extname" do
      outfile = UnixUtils.perl(@infile, 's/bad/good/g')
      File.extname(outfile).must_equal File.extname(@infile)
      safe_delete outfile
    end
  end

  describe :awk do
    before do
      @f = Tempfile.new('awk.txt')
      @f.write "badWord\n"*10_000
      @f.flush
      @infile = @f.path
    end
    after do
      @f.close
    end
    it "processes a file" do
      outfile = UnixUtils.awk(@infile, '{gsub(/bad/, "good"); print}')
      File.read(outfile).must_equal "goodWord\n"*10_000
      safe_delete outfile
    end
    it "does not touch the infile" do
      assert_does_not_touch :awk, @infile, '{gsub(/bad/, "good"); print}'
    end
    it "keeps the original extname" do
      outfile = UnixUtils.awk(@infile, '{gsub(/bad/, "good"); print}')
      File.extname(outfile).must_equal File.extname(@infile)
      safe_delete outfile
    end
  end

  describe :unix2dos do
    before do
      @f = Tempfile.new('unix2dos.txt')
      @f.write "unix\n"*5_000
      @f.write "dos\r\n"*5_000
      @f.flush
      @infile = @f.path
    end
    after do
      @f.close
    end
    it 'converts newlines' do
      outfile = UnixUtils.unix2dos @infile
      File.read(outfile).must_equal("unix\r\n"*5_000 + "dos\r\n"*5_000)
      safe_delete outfile
    end
  end

  describe :dos2unix do
    before do
      @f = Tempfile.new('dos2unix.txt')
      @f.write "dos\r\n"*5
      @f.write "unix\n"*5
      @f.flush
      @infile = @f.path
    end
    after do
      @f.close
    end
    it 'converts newlines' do
      outfile = UnixUtils.dos2unix @infile
      File.read(outfile).must_equal("dos\n"*5 + "unix\n"*5)
      safe_delete outfile
    end
  end

  describe :wc do
    before do
      @f = Tempfile.new('wc.txt')
      @f.write "dos line\r\n"*50_000
      @f.write "unix line\n"*50_000
      @f.flush
      @infile = @f.path
    end
    after do
      @f.close
    end
    it 'counts lines, words, and bytes' do
      UnixUtils.wc(@infile).must_equal [50_000+50_000, 100_000+100_000, 500_000+500_000]
    end
  end

  describe :sed do
    before do
      @f = Tempfile.new('sed.txt')
      @f.write "badWord\n"*10_000
      @f.flush
      @infile = @f.path
    end
    after do
      @f.close
    end
    it "processes a file" do
      outfile = UnixUtils.sed(@infile, 's/bad/good/g')
      File.read(outfile).must_equal "goodWord\n"*10_000
      safe_delete outfile
    end
    it "does not touch the infile" do
      assert_does_not_touch :sed, @infile, 's/bad/good/g'
    end
    it "keeps the original extname" do
      outfile = UnixUtils.sed(@infile, 's/bad/good/g')
      File.extname(outfile).must_equal File.extname(@infile)
      safe_delete outfile
    end

  end

  describe :tail do
    before do
      @a2z = ('a'..'z').to_a
      @f = Tempfile.new('tail.txt')
      @f.write @a2z.join("\n")
      @f.flush
      @infile = @f.path
    end
    after do
      @f.close
    end
    it 'gets last three lines' do
      outfile = UnixUtils.tail(@infile, 3)
      File.read(outfile).must_equal @a2z.last(3).join("\n")
      safe_delete outfile
    end
    it 'gets trailing lines starting with the third line (inclusive)' do
      outfile = UnixUtils.tail(@infile, '+3')
      File.read(outfile).must_equal @a2z[2..-1].join("\n")
      safe_delete outfile
    end
    it 'has a related tail_s method' do
      str = UnixUtils.tail_s(@infile, 3)
      str.must_equal @a2z.last(3).join("\n")
    end
  end

  describe :head do
    before do
      @a2z = ('a'..'z').to_a
      @f = Tempfile.new('head.txt')
      @f.write @a2z.join("\n")
      @f.flush
      @infile = @f.path
    end
    after do
      @f.close
    end
    it 'gets first three lines' do
      outfile = UnixUtils.head(@infile, 3)
      File.read(outfile).must_equal(@a2z.first(3).join("\n") + "\n")
      safe_delete outfile
    end
    it 'has a related head_s method' do
      str = UnixUtils.head_s(@infile, 3)
      str.must_equal(@a2z.first(3).join("\n") + "\n")
    end
  end

  describe :cut do
    before do
      @a2z = ('a'..'z').to_a
      @f = Tempfile.new('cut.txt')
      10.times do
        @f.write(@a2z.join + "\n")
      end
      @f.flush
      @infile = @f.path
    end
    after do
      @f.close
    end
    it 'cuts out character positions' do
      outfile = UnixUtils.cut(@infile, '1,12,13,15,19,20')
      almosts = (0..9).map { |i| 'almost' }.join("\n") + "\n"
      File.read(outfile).must_equal almosts
      safe_delete outfile
    end
    it 'cuts out character ranges (inclusive)' do
      outfile = UnixUtils.cut(@infile, '3-6')
      cdefs = (0..9).map { |i| 'cdef' }.join("\n") + "\n"
      File.read(outfile).must_equal cdefs
      safe_delete outfile
    end
  end

  describe :iconv do
    it 'converts files from utf-8 to latin1' do
      outfile = UnixUtils.iconv('utf8.txt', 'ISO-8859-1', 'UTF-8')
      UnixUtils.md5sum(outfile).must_equal UnixUtils.md5sum('iso-8859-1.txt')
      safe_delete outfile
    end
    it 'converts files from latin1 to utf-8' do
      outfile = UnixUtils.iconv('iso-8859-1.txt', 'UTF-8', 'ISO-8859-1')
      UnixUtils.md5sum(outfile).must_equal UnixUtils.md5sum('utf8.txt')
      safe_delete outfile
    end
  end

  describe :tmp_path do
    it "includes basename of ancestor" do
      UnixUtils.tmp_path("dirname1/dirname2/basename.extname").must_include 'basename'
    end
    it "includes extname of ancestor" do
      UnixUtils.tmp_path("dirname1/dirname2/basename.extname").must_include 'extname'
    end
    it "optionally appends extname" do
      File.extname(UnixUtils.tmp_path("dirname1/dirname2/basename.extname", '.foobar')).must_equal '.foobar'
    end
    it "doesn't create excessively long filenames" do
      100.times { File.basename(UnixUtils.tmp_path("a"*5000)).length.must_equal(255) }
    end
    it "doesn't include directory part of ancestor" do
      UnixUtils.tmp_path("dirname1/dirname2/basename.extname").wont_include 'dirname1'
    end
    it "includes unix_utils part only once" do
      one = UnixUtils.tmp_path('basename.extname')
      File.basename(one).start_with?('unix_utils').must_equal true
      one.scan(/unix_utils/).length.must_equal 1
      again = UnixUtils.tmp_path(one)
      File.basename(again).start_with?('unix_utils').must_equal true
      again.scan(/unix_utils/).length.must_equal 1
      and_again = UnixUtils.tmp_path(again)
      File.basename(and_again).start_with?('unix_utils').must_equal true
      and_again.scan(/unix_utils/).length.must_equal 1
      and_again.must_include('basename.extname')
    end
  end

  # not really for public consumption
  describe :spawn do
    before do
      @f = Tempfile.new('spawn.txt')
      @f.write "dos line\r\n"*50_000
      @f.write "unix line\n"*50_000
      @f.flush
      @infile = @f.path
    end
    after do
      @f.close
    end
    it 'reads and writes everything' do
      wc_output = UnixUtils.spawn(['wc'], :read_from => @infile)
      wc_output.strip.split(/\s+/)[0..2].map { |s| s.to_i }.must_equal [50_000+50_000, 100_000+100_000, 500_000+500_000]
    end
  end
end


test/regression/test_sed_large_file.rb
# encoding: UTF-8
require 'helper'
require 'tempfile'

describe "handle large file with sed" do
  let(:normal_line_no)  { 124000 }
  let(:invalid_line_no) { 10 }
  let(:input_file)      { Tempfile.open('test', '/tmp') }

  before do
    begin
      (1..normal_line_no).each do |i|
        input_file.puts "\"psmith01\",\"CLASS2B\",\"Peter Smith 1\",\"YEAR2\",\"1\",\"N\",\"ADVANCED\",\"STAFF\",\"1\",\"Y\",\"Y\",\"psmith01\",\"CLASS2B\",\"Peter Smith 1\",\"YEAR2\",\"1\",\"N\",\"ADVANCED\",\"STAFF\",\"1\",\"Y\",\"Y\",\"psmith01\",\"CLASS2B\",\"Peter Smith 1\",\"YEAR2\",\"1\",\"N\",\"ADVANCED\",\"STAFF\",\"1\",\"Y\",\"Y\",\"psmith01\",\"CLASS2B\",\"Peter Smith 1\",\"YEAR2\",\"1\",\"N\",\"ADVANCED\",\"STAFF\",\"1\",\"Y\",\"Y\""
      end
      (1..invalid_line_no).each do |i|
        input_file.puts "@!!@"
      end
      input_file.flush
    ensure
      input_file.close
    end
  end

  it "should have 124000 lines in output file" do
    output_path = UnixUtils.sed(input_file.path, ':a', "1,#{invalid_line_no}!{P;N;D;};N;ba")
    UnixUtils.wc(output_path).first.must_equal normal_line_no
  end

  after do
    input_file.unlink
  end
end


lib/unix_utils.rb
require 'fileutils'
require 'tmpdir'
require 'uri'
require 'stringio'
require 'open3'
require 'securerandom'
require "unix_utils/version"

module UnixUtils

  BUFSIZE = 2**16

  def self.curl(url, form_data = nil)
    outfile = tmp_path url
    if url.start_with?('file://') or not url.include?('://')
      # deal with local files
      infile = File.expand_path url.sub('file://', '')
      unless File.readable?(infile)
        raise "[unix_utils] #{url.inspect} does not exist or is not readable on the local filesystem."
      end
      FileUtils.cp infile, outfile
      return outfile
    end
    uri = URI.parse url
    argv = [ 'curl', '--location', '--show-error', '--silent', '--compressed', '--header', 'Expect: ' ]
    if form_data
      argv += [ '--data', form_data ]
    end
    argv += [ uri.to_s, '--output', outfile ]
    spawn argv
    outfile
  end

  #--
  # most platforms
  # $ openssl dgst -sha256 .bash_profile
  # SHA256(.bash_profile)= ae12206aaa35dc96273ed421f4e85ca26a1707455e3cc9f054c7f5e2e9c53df6
  # ubuntu 11.04
  # $ shasum -a 256 --portable .mysql_history
  # 856aa27deb0b80b41031c2ddf722af28ba2a8c4999ff9cf2d45f33bc67d992ba ?.mysql_history
  # fedora 7
  # $ sha256sum --binary .bash_profile
  # 01b1210962b3d1e5e1ccba26f93d98efbb7b315b463f9f6bdb40ab496728d886 *.bash_profile
  def self.shasum(infile, algorithm)
    infile = File.expand_path infile
    if available?('shasum')
      argv = ['shasum', '--binary', '-a', algorithm.to_s, infile]
      stdout = spawn argv
      stdout.strip.split(' ').first
    else
      argv = ['openssl', 'dgst', "-sha#{algorithm}", infile]
      stdout = spawn argv
      stdout.strip.split(' ').last
    end
  end

  #--
  # os x 10.6.8; most platforms
  # $ openssl dgst -md5 .bashrc
  # MD5(.bashrc)= 88f464fb6d1d6fe9141135248bf7b265
  # ubuntu 11.04; fedora 7; gentoo
  # $ md5sum --binary .mysql_history
  # 8d01e54ab8142d6786850e22d55a1b6c *.mysql_history
  def self.md5sum(infile)
    infile = File.expand_path infile
    if available?('md5sum')
      argv = ['md5sum', '--binary', infile]
      stdout = spawn argv
      stdout.strip.split(' ').first
    else
      argv = ['openssl', 'dgst', '-md5', infile]
      stdout = spawn argv
      stdout.strip.split(' ').last
    end
  end

  def self.du(srcdir)
    srcdir = File.expand_path srcdir
    argv = ['du', '-sk', srcdir]
    stdout = spawn argv
    stdout.strip.split(/\s+/).first.to_i
  end

  def self.wc(infile)
    infile = File.expand_path infile
    argv = ['wc', infile]
    stdout = spawn argv
    stdout.strip.split(/\s+/)[0..2].map { |s| s.to_i }
  end

  # --

  def self.unzip(infile)
    infile = File.expand_path infile
    destdir = tmp_path infile
    FileUtils.mkdir destdir
    argv = ['unzip', '-qq', '-n', infile, '-d', destdir]
    spawn argv
    destdir
  end

  def self.untar(infile)
    infile = File.expand_path infile
    destdir = tmp_path infile
    FileUtils.mkdir destdir
    argv = ['tar', '-xf', infile, '-C', destdir]
    spawn argv
    destdir
  end

  def self.gunzip(infile)
    infile = File.expand_path infile
    outfile = tmp_path infile
    argv = ['gunzip', '--stdout', infile]
    spawn argv, :write_to => outfile
    outfile
  end

  def self.bunzip2(infile)
    infile = File.expand_path infile
    outfile = tmp_path infile
    argv = ['bunzip2', '--stdout', infile]
    spawn argv, :write_to => outfile
    outfile
  end

  # --

  def self.bzip2(infile)
    infile = File.expand_path infile
    outfile = tmp_path infile, '.bz2'
    argv = ['bzip2', '--keep', '--stdout', infile]
    spawn argv, :write_to => outfile
    outfile
  end

  def self.tar(srcdir)
    srcdir = File.expand_path srcdir
    outfile = tmp_path srcdir, '.tar'
    argv = ['tar', '-cf', outfile, '-C', srcdir, '.']
    spawn argv
    outfile
  end

  def self.zip(srcdir)
    srcdir = File.expand_path srcdir
    outfile = tmp_path srcdir, '.zip'
    argv = ['zip', '-rq', outfile, '.']
    spawn argv, :chdir => srcdir
    outfile
  end

  def self.gzip(infile)
    infile = File.expand_path infile
    outfile = tmp_path infile, '.gz'
    argv = ['gzip', '--stdout', infile]
    spawn argv, :write_to => outfile
    outfile
  end

  # --

  def self.awk(infile, *expr)
    infile = File.expand_path infile
    outfile = tmp_path infile
    bin = available?('gawk') ? 'gawk' : 'awk'
    argv = [bin, expr, infile].flatten
    spawn argv, :write_to => outfile
    outfile
  end

  # Yes, this is a very limited use of perl.
  def self.perl(infile, *expr)
    infile = File.expand_path infile
    outfile = tmp_path infile
    argv = [ 'perl', expr.map { |e| ['-pe', e] }, infile ].flatten
    spawn argv, :write_to => outfile
    outfile
  end

  def self.unix2dos(infile)
    infile = File.expand_path infile
    if available?('gawk') or available?('awk')
      awk infile, '{ sub(/\r/, ""); printf("%s\r\n", $0) }'
    else
      perl infile, 's/\r\n|\n|\r/\r\n/g'
    end
  end

  def self.dos2unix(infile)
    infile = File.expand_path infile
    if available?('gawk') or available?('awk')
      awk infile, '{ sub(/\r/, ""); printf("%s\n", $0) }'
    else
      perl infile, 's/\r\n|\n|\r/\n/g'
    end
  end

  # POSIX sed, whether it's provided by sed or gsed
  def self.sed(infile, *expr)
    infile = File.expand_path infile
    outfile = tmp_path infile
    bin = available?('sed') ? 'sed' : ['gsed', '--posix']
    argv = [ bin, expr.map { |e| ['-e', e] }, infile ].flatten
    spawn argv, :write_to => outfile
    outfile
  end

  def self.tail(infile, lines)
    infile = File.expand_path infile
    outfile = tmp_path infile
    argv = ['tail', '-n', lines.to_s, infile]
    spawn argv, :write_to => outfile
    outfile
  end

  def self.head(infile, lines)
    infile = File.expand_path infile
    outfile = tmp_path infile
    argv = ['head', '-n', lines.to_s, infile]
    spawn argv, :write_to => outfile
    outfile
  end

  # specify character_positions as a string like "3-5" or "3,9-10"
  def self.cut(infile, character_positions)
    infile = File.expand_path infile
    outfile = tmp_path infile
    argv = ['cut', '-c', character_positions, infile]
    spawn argv, :write_to => outfile
    outfile
  end

  def self.iconv(infile, to, from)
    infile = File.expand_path infile
    outfile = tmp_path infile
    argv = ['iconv', '-c', '-t', to, '-f', from, infile]
    spawn argv, :write_to => outfile
    outfile
  end

  def self.available?(bin) # :nodoc:
    bin = bin.to_s
    return @@available_query[bin] if defined?(@@available_query) and @@available_query.is_a?(Hash) and @@available_query.has_key?(bin)
    @@available_query ||= {}
    `which #{bin}`
    @@available_query[bin] = $?.success?
  end

  def self.tmp_path(ancestor, extname = nil) # :nodoc:
    ancestor = ancestor.to_s
    extname ||= File.extname ancestor
    basename = File.basename ancestor, extname
    basename.gsub! /^unix_utils_[a-f0-9]{8,}_/, ''
    basename.gsub! /\W+/, '_'
    File.join Dir.tmpdir, "unix_utils_#{SecureRandom.hex(4)}_#{basename[0..(234-extname.length)]}#{extname}"
  end

  def self.spawn(argv, options = {}) # :nodoc:
    input = if (read_from = options[:read_from])
      if RUBY_DESCRIPTION =~ /jruby 1.7.0/
        raise "[unix_utils] Can't use `#{argv.first}` since JRuby 1.7.0 has a broken IO implementation!"
      end
      File.open(read_from, 'r')
    end
    output = if (write_to = options[:write_to])
      output_redirected = true
      File.open(write_to, 'wb')
    else
      output_redirected = false
      StringIO.new
    end
    error = StringIO.new
    if (chdir = options[:chdir])
      Dir.chdir(chdir) do
        _spawn argv, input, output, error
      end
    else
      _spawn argv, input, output, error
    end
    error.rewind
    unless (whole_error = error.read).empty?
      $stderr.puts "[unix_utils] `#{argv.join(' ')}` STDERR:"
      $stderr.puts whole_error
    end
    unless output_redirected
      output.rewind
      output.read
    end
  ensure
    [input, output, error].each { |io| io.close if io and not io.closed? }
  end

  def self._spawn(argv, input, output, error)
    # lifted from posix-spawn
    # https://github.com/rtomayko/posix-spawn/blob/master/lib/posix/spawn/child.rb
    Open3.popen3(*argv) do |stdin, stdout, stderr|
      readers = [stdout, stderr]
      if RUBY_DESCRIPTION =~ /jruby 1.7.0/
        readers.delete stderr
      end
      writers = if input
        [stdin]
      else
        stdin.close
        []
      end
      while readers.any? or writers.any?
        ready = IO.select(readers, writers, readers + writers)
        # write to stdin stream
        ready[1].each do |fd|
          begin
            boom = nil
            size = fd.write input.read(BUFSIZE)
          rescue Errno::EPIPE => boom
          rescue Errno::EAGAIN, Errno::EINTR
          end
          if boom || size < BUFSIZE
            stdin.close
            input.close
            writers.delete stdin
          end
        end
        # read from stdout and stderr streams
        ready[0].each do |fd|
          buf = (fd == stdout) ? output : error
          if fd.eof?
            readers.delete fd
            fd.close
          else
            begin
              buf << fd.readpartial(BUFSIZE)
            rescue Errno::EAGAIN, Errno::EINTR
            end
          end
        end
      end
      # thanks @tmm1 and @rtomayko for showing how it's done!
    end
  end

  def self.method_missing(method_id, *args)
    base_method_id = method_id.to_s.chomp('_s')
    if respond_to?(base_method_id)
      begin
        outfile = send(*([base_method_id]+args))
        File.read outfile
      ensure
        FileUtils.rm_f outfile
      end
    else
      super
    end
  end
end
